lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -94,14 +94,215 @@ class _AwaitVisitor extends RecursiveAstVisitor {
+    // Stop visiting when we arrive at a function body.
+    // Awaits inside it don't matter.
+  }
+}
+
+/// An enum of two values which describe the presence of a "mounted check."
+///
+/// A mounted check is a check of whether a bool-typed identifier, 'mounted',
+/// is checked to be `true` or `false`, in a position which affects control
+/// flow.
+enum _MountedCheck {
+  positive,
+  negative;
+
+  _MountedCheck get negate => switch (this) {
+        _MountedCheck.positive => _MountedCheck.negative,
+        _MountedCheck.negative => _MountedCheck.positive,
+      };
+}
+
+/// A visitor whose `visit*` methods return whether a "mounted check" is found
+/// which properly guards [child].
+///
+/// The entrypoint for this visitor is [AstNodeExtension.isMountedCheckFor].
+///
+/// A mounted check "guards" [child] if control flow can only reach [child] if
+/// 'mounted' is `true`. Such checks can take many forms:
+///
+/// * A mounted check in an if-condition can be a simple guard for nodes in the
+///   if's then-statement or the if's else-statement, depending on the polarity
+///   of the check. So `if (mounted) { child; }` has a proper mounted check and
+///   `if (!mounted) {} else { child; }` has a proper mounted check.
+/// * A statement in a series of statements containing a mounted check can guard
+///   the later statements if control flow definitely exits in the case of a
+///   `false` value for 'mounted'. So `if (mounted) { return; } child;` has a
+///   proper mounted check.
+/// * A mounted check in a try-statement can only guard later statements if it
+///   is found in the `finally` section, as no statements found in the `try`
+///   section or any `catch` sections are not guaranteed to have run before the
+///   later statements.
+/// * etc.
+///
+/// Each `visit*` method can return one of three values:
+/// * `null` means the node does not guard [child] with a mounted check.
+/// * [_MountedCheck.positive] means the node guards [child] with a positive
+///   mounted check.
+/// * [_MountedCheck.negative] means the node guards [child] with a negative
+///   mounted check.
+class _MountedCheckVisitor extends SimpleAstVisitor<_MountedCheck> {
+  static const mountedName = 'mounted';
+
+  final AstNode child;
+
+  _MountedCheckVisitor({required this.child});
+
+  @override
+  _MountedCheck? visitBinaryExpression(BinaryExpression node) {
+    // TODO(srawlins): Currently this method doesn't take `child` into account;
+    // it assumes `child` is part of a statement that follows this expression.
+    // We need to account for `child` being an actual descendent of `node` in
+    // order to properly handle code like
+    // * `if (mounted || child)`,
+    // * `if (!mounted && child)`,
+    // * `if (mounted || (condition && child))`,
+    // * `if ((mounted || condition) && child)`, etc.
+    if (node.isAnd) {
+      return node.leftOperand.accept(this) ?? node.rightOperand.accept(this);
```

Comment:
Right, I want to keep this a `SimpleAstVisitor` so that the visitations are only performed through explicit `visit*` methods here.

I think it's clean enough to require visiting every possible node. The previous implementation, `bool isMountedCheck()` has this same requirement (and today does not handle parenthesized expressions).